editoast/src/modelsv2/database/connection_pool.rs

```diff
@@ -0,0 +1,275 @@
+    ///
+    /// # Test mode
+    ///
+    /// In test mode, this function will return the same connection that has been setup to drop all
+    /// modifications once the test ends. This connection is intended to be used in unit tests to
+    /// ensure the isolation of each test.
+    ///
+    /// ## Pitfalls
+    ///
+    /// However, this comes with several limitations on how connections are used globally.
+    ///
+    /// 1. Once a connection is used, it should be dropped **AS SOON AS POSSIBLE**. Failing to do so
+    ///    may lead to deadlocks. Example:
+    ///
+    /// ```rust
+    /// let conn = pool.get_ok();
+    /// // Do something with conn
+    ///
+    /// // This will deadlock because `conn` hasn't been droped yet. Since this function is
+    /// // not async, this is equivalent to an infinite loop.
+    /// let conn2 = pool.get_ok();
+    /// ```
+    ///
+    /// 2. If several futures are spawned and each use their own connection, you should make sure
+    ///    that the connection usage follows its acquisition. Failing to do so is equivalent to
+    ///    the following example:
+    ///
+    /// ```rust
+    /// let conn_futures = (0..10).map(|_| async { pool.get() });
+    /// let deadlock = futures::future::join_all(conn_futures).await;
+    /// ```
+    ///
+    /// ## Guidelines
+    ///
+    /// To prevent these issues, prefer the following patterns:
+    ///
+    /// - Don't declare a variable for a single-use connection:
+    ///
+    /// ```rust
+    /// // do
+    /// my_function_using_conn(pool.get().await?.deref_mut()).await;
+    /// // instead of
+    /// let conn = &mut pool.get().await?;
+    /// my_function_using_conn(conn).await;
+    /// ```
+    ///
+    /// - If a connection is used repeatedly, prefer using explicit scoping:
+    ///
+    /// ```rust
+    /// let my_results = {
+    ///     let conn = &mut pool.get().await?;
+    ///     foo(conn).await + bar(conn).await
+    /// };
+    /// // you may acquire a new connection afterwards
+    /// ```
+    ///
+    /// - If you need to open several connections, then the connection must be
+    ///   acquired just before its usage, and dropped just after, **all in the same future**.
+    ///   And these fututres must all be awaited before attempting to acquire a new connection.
+    ///
+    /// ```rust
+    /// let operations =
+    ///     items.into_iter()
+    ///         .zip(pool.iter_conn())
+    ///         .map(|(item, conn)| async {
+    ///             let conn = conn.await?; // note the await here
+    ///             item.do_something(conn).await
+    ///         });
+    /// let results = futures::future::try_join_all(operations).await?;
+    /// // you may acquire a new connection afterwards
+    /// ```
+    pub async fn get(&self) -> Result<DbConnectionV2, DbConnectionError> {
+        #[cfg(test)]
+        let conn = self.get_test().await;
+        #[cfg(not(test))]
+        let conn = self.get_proxy().await;
```

Comment:
Why not naming `get_test` and `get_proxy` with the same name (`get_connection`?) to avoid this `#[cfg()]` branch mechanism.

```suggestion
        let conn = self.get_proxy().await;
```

Reply:
I think this is for the documentation. We are not sure which version of the doc editor will be chosen, the test or non-test version. By having only one function, this is not a problem.

Reply:
I do not understand, `get_proxy` and `get_test` don't have a documentation.

Reply:
Yes, because they are private functions. If we use cfg(test) or not(test) with two different functions having the same name, which version of the documentation will be shown in the editor? I guess it depends on the configuration of each editor. With only one function without any cfg and proper documentation, we can be sure there are no problems with that.

Reply:
The documentation covers the behavior of the function with in both `test` and `not(test)`. This is intentional since the `not(test)` usage of the connection is limited by how testing works (e.g. no `join!` on connection acquisition).

The problem is with `rust-analyzer` that has `test` on by default, so when we hover a function, its `test` documentation will be displayed. I felt it was weird to document only the `#[cfg(test)] fn get` but I also didn't want to duplicate the whole documentation.

So the `get` function serves as a single entry point, to have the documentation in a single place, and then the branching occurs.